src/eventProcessor/GoogleAnalyticsEventProcessor.js

```diff
@@ -0,0 +1,155 @@
+ * POST body instead of as event parameters.
+ * @const {!Object<string,boolean>}
+ */
+const TOP_LEVEL_PARAMS = {
+  'client_id': true,
+  'user_id': true,
+  'timestamp_micros': true,
+  'user_properties': true,
+  'non_personalized_ads': true,
+};
+
+const DEFAULT_MEASUREMENT_URL = 'https://www.google-analytics.com/mp/collect';
+
+/**
+ * A class that processes events pushed to the data layer
+ * by constructing and sending Google Analytics events via
+ * Measurement Protocol (App + Web).
+ * For its Google Analytic events, it generates a unique client ID
+ * and stores it in the long term storage model provided.
+ * @implements {EventProcessor}
+ */
+class GoogleAnalyticsEventProcessor {
+  /**
+   * Constructs a Google Analytics Event Processor using the following
+   * optional arguments within the options obejct:
+   *
+   * api_secret: Google Analytics API Secret. If provided, will be attached to
+   *     query parameter of the same name when sending events.
+   * measurement_id: Property to be measured Google Analytics Identifier. If
+   *     provided, will be attached to query parameter of the same name when
+   *     sending events.
+   * measurement_url: URL endpoint to send events to. Defaults to
+   *     Google Analytics collection endpoint.
+   * client_id_expires: Number of seconds to store the client ID in long term
+   *     storage. Defaults to two years.
+   * automatic_params: Array of event parameters that will be searched for in
+   *     the global data model and pulled into all events if found.
+   * @param {{
+   *     'api_secret': (string|undefined),
+   *     'measurement_id': (string|undefined),
+   *     'measurement_url': string,
+   *     'client_id_expires': number,
+   *     'automatic_params': !Array<string>,
+   * }=} optionsObject
+   */
+  constructor({
+      'api_secret': apiSecret,
+      'measurement_id': measurementId,
+      'measurement_url': measurementUrl = DEFAULT_MEASUREMENT_URL,
+      'client_id_expires': clientIdExpires = 2 * 365 * 24 * 60 * 60,
+      'automatic_params': userAutomaticParams = [],
+    } = {}) {
+    // TODO(kjgalvan):: log error if measurementUrl is default value and
+    // either apiSecret or measurementId are undefined.
+
+    /**
+     * Parameters that are important to all events and will be searched for
+     * globally in the data model.
+     * @private @const {!Object<string, boolean>}
+     */
+    this.automaticParams_ = {
+      'page_path': true,
+      'page_location': true,
+      'page_title': true,
+      'user_id': true,
+      'client_id': true,
+    };
+
+    // Add user provided params to automatic param list
+    for (let i = 0; i < userAutomaticParams.length; ++i) {
+      this.automaticParams_[userAutomaticParams[i]] = true;
+    }
+
+    /**
+     * An API secret that can be generated via Google Analytics UI
+     * NOTE: This processor makes public network requests using this
+     * API secret. To avoid spam, regenerate API secrets regularly.
```

Comment:
No action needed in this request, but this note should be made in much more visible documentation for the processor as well!